### PR TITLE
Fix constraint violation when deleting a book

### PR DESCRIPTION
**Pull request by amit1307**
## Summary of change

Fix the issue when try to delete a book 

## Related issue

https://github.com/knjk04/book-project/issues/114

### CODE REVIEW

**Comment by amit1307** [comment]
The code coverage reduction is due the fact that change was in domain class and we don't normally have tests for domain classes. But to test the flow impacted by the change in domain class, 2 new tests have been added. Since the new tests are invoking JPA's methods so I am guessing the code coverage is not increased by the new tests.

**Comment by amit1307** [comment]
> Thanks for this.
> 
> I left some small suggestions for you to address. Also, was there a particular reason that you decided to test the repository rather than the service?
> 
> Once you've seen to the changes, could you rebase? You should then get the patch for #269

The only reason that I had in mind for testing repo instead of service was the nature of the issue. Since the issue is purely related to delete function so I felt repo test would be more appropriate.

File: src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    void shouldSuccessfullyDeleteABook_whenAuthorHasNoOtherBooks() {
+        // given
+        Author author = authorRepository.saveAndFlush(new Author("firstName", "lastName"));
+        Book book = bookRepository.save(new Book("someTitle", author, predefinedShelf));
+
+        // when
+        bookRepository.delete(book);
+
+        // then
+        Assertions.assertEquals(0, bookRepository.findAll().size());
```
[the remark this answers is not in the document]
Comment (amit1307) [reply src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java]: Done

File: src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        Author author = authorRepository.saveAndFlush(new Author("firstName", "lastName"));
+        Book book1 = bookRepository.save(new Book("someTitle", author, predefinedShelf));
+        Book book2 = bookRepository.save(new Book("someOtherTitle", author, predefinedShelf));
+
+        // when
+        bookRepository.delete(book1);
+
+        // then
+        Assertions.assertEquals(1, bookRepository.findAll().size());
+        Assertions.assertSame(bookRepository.findAll().get(0), book2);
```
[the remark this answers is not in the document]
Comment (amit1307) [reply src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java:46]: Done

File: src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    @Autowired
+    private AuthorRepository authorRepository;
+
+    @Autowired
+    private BookRepository bookRepository;
+
+    PredefinedShelf predefinedShelf;
+
+    @BeforeEach
+    void setUp() {
```
[the remark this answers is not in the document]
Comment (amit1307) [reply src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java]: Done

[the remark this answers is not in the document]
Comment (amit1307) [reply src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java]: Just curious  if this is your personal preference or is there any standard recommendation for not using this method in camelCase?

[the remark this answers is not in the document]
Comment (knjk04) [reply src/test/java/com/karankumar/bookproject/backend/repository/BookRepositoryTest.java]: Personal preference. It's fine to style it as two words, but we just want to remain consistent (we've been styling it as one word elsewhere)

**Comment by knjk04** [comment]
OK, thanks

**Comment by knjk04** [comment]
Could you please rebase? Once you've done that, and providing it still works, it should be ready to be merged in

**Comment by amit1307** [comment]
> Could you please rebase? Once you've done that, and providing it still works, it should be ready to be merged in

Yes I have a quick testing and it works.

**Comment by knjk04** [comment]
I saw your message on the issue description. I added a workaround for the #245 bug, which was partially blocking this, so could you still rebase? I think I recently added it; it's not in your branch.

**Comment by amit1307** [comment]
> I saw your message on the issue description. I added a workaround for the #245 bug, which was partially blocking this, so could you still rebase? I think I recently added it; it's not in your branch.

sure, have rebased and pushed.

**Review by knjk04 — APPROVED** [review]
Thanks a lot for this!

**Pull request merged**